runtime/src/main/java/com/dylibso/chicory/runtime/Instance.java

```diff
@@ -433,4 +455,512 @@ public void onExecution(Instruction instruction, long[] operands, MStack stack)
+                        for (int j = 0; j < cnt; j++) {
+                            HostMemory m = hostImports.memory(j);
+                            if (checkName.apply(m)) {
+                                validateHostMemoryType((MemoryImport) i, m);
+                                hostMems[hostMemIdx] = m;
+                                found = true;
+                                break;
+                            }
+                        }
+                        hostMemIdx++;
+                        break;
+                    case TABLE:
+                        cnt = hostImports.tableCount();
+                        for (int j = 0; j < cnt; j++) {
+                            HostTable t = hostImports.table(j);
+                            if (checkName.apply(t)) {
+                                validateHostTableType((TableImport) i, t);
+                                hostTables[hostTableIdx] = t;
+                                found = true;
+                                break;
+                            }
+                        }
+                        hostTableIdx++;
+                        break;
+                }
+                if (!found) {
+                    if (importValidation) {
+                        throw new UnlinkableException(
+                                "unknown import, could not find host function for import number: "
+                                        + impIdx
+                                        + " named "
+                                        + name);
+                    } else {
+                        System.err.println(
+                                "Could not find host function for import number: "
+                                        + impIdx
+                                        + " named "
+                                        + name);
+                    }
+                }
+            }
+
+            var result = new HostImports(hostFuncs, hostGlobals, hostMems, hostTables);
+            return result;
+        }
+
+        private Map<String, Export> genExports(ExportSection export) {
+            var exports = new HashMap<String, Export>();
+            int cnt = export.exportCount();
+            for (int i = 0; i < cnt; i++) {
+                Export e = export.getExport(i);
+                if (exports.containsKey(e.name())) {
+                    throw new InvalidException("duplicate export name " + e.name());
+                }
+                exports.put(e.name(), e);
+            }
+            return exports;
+        }
+
+        public Instance build() {
+            validateModule(module);
+            Map<String, Export> exports = genExports(module.exportSection());
+            var globalInitializers = module.globalSection().globals();
+
+            var dataSegments = module.dataSection().dataSegments();
+
+            // TODO i guess we should explode if this is the case, is this possible?
```

Comment:
I'd recommend any time we have one of these situations to just assert it. Catching it is always worth the assertion cost.

Reply:
This comment is floating around since ever, I think it doesn't apply anymore, I remove it.